sycl/include/CL/sycl/id.hpp

```diff
@@ -119,22 +120,119 @@ template <int dimensions = 1> class id : public detail::array<dimensions> {
 
-  #undef __SYCL_GEN_OPT
+#undef __SYCL_GEN_OPT
 
-  // OP is: +=, -=, *=, /=, %=, <<=, >>=, &=, |=, ^=
-  #define __SYCL_GEN_OPT(op)                                                   \
-    id<dimensions> &operator op(const id<dimensions> &rhs) {                   \
-      for (int i = 0; i < dimensions; ++i) {                                   \
-        this->common_array[i] op rhs.common_array[i];                          \
-      }                                                                        \
-      return *this;                                                            \
+// OP is: +=, -=, *=, /=, %=, <<=, >>=, &=, |=, ^=
+#define __SYCL_GEN_OPT(op)                                                     \
+  id<dimensions> &operator op(const id<dimensions> &rhs) {                     \
+    for (int i = 0; i < dimensions; ++i) {                                     \
+      this->common_array[i] op rhs.common_array[i];                            \
     }                                                                          \
-    id<dimensions> &operator op(const size_t &rhs) {                           \
-      for (int i = 0; i < dimensions; ++i) {                                   \
-        this->common_array[i] op rhs;                                          \
-      }                                                                        \
-      return *this;                                                            \
+    return *this;                                                              \
+  }                                                                            \
+  id<dimensions> &operator op(const size_t &rhs) {                             \
+    for (int i = 0; i < dimensions; ++i) {                                     \
+      this->common_array[i] op rhs;                                            \
     }                                                                          \
+    return *this;                                                              \
+  }
+
+  __SYCL_GEN_OPT(+=)
+  __SYCL_GEN_OPT(-=)
+  __SYCL_GEN_OPT(*=)
+  __SYCL_GEN_OPT(/=)
+  __SYCL_GEN_OPT(%=)
+  __SYCL_GEN_OPT(<<=)
+  __SYCL_GEN_OPT(>>=)
+  __SYCL_GEN_OPT(&=)
+  __SYCL_GEN_OPT(|=)
+  __SYCL_GEN_OPT(^=)
+
+#undef __SYCL_GEN_OPT
+};
+
+template <> class id<1> : public detail::array<1> {
+private:
+  using base = detail::array<1>;
+
+public:
+  id() = default;
+
+  /* The following constructor is only available in the id struct
+   * specialization where: dimensions==1 */
+  id<1>(size_t dim0) : base(dim0) {}
+
+  id<1>(const range<1> &range_size) : base(range_size.get(0)) {}
+
+  template <bool with_offset = true>
+  id<1>(const item<1, with_offset> &item) : base(item.get_id(0)) {}
+
+  explicit operator range<1>() const {
+    range<1> result(detail::InitializedVal<1, range>::template get<0>());
+    result[0] = this->get(0);
+    return result;
+  }
+
+  operator size_t() const { return this->get(0); }
```

Comment:
We use SFINAE to enable methods for specific template parameter value. It should allow you to avoid a lot of code duplication. Right?
Why do you decide to specialize the whole class?

Reply:
SFINAE on conversion operators is not a good idea (on anything other than conversion operators it is totally fine). For conversion operators it changes the behavior. In this case the int conversion after the implicit conversion doesn't work anymore (if only size_t is allowed by SFINAE) or it is ambiguous (if all int-types are allowed by SFINAE). See https://gitlab.devtools.intel.com/SYCL/features/issues/20#note_2368711.

Reply:
@bader, 
In case of template implicit conversion (SFINAE) it does not carry out two or more conversions. For example, in case of template implicit conversion the next code will work:
`size_t a = id<1>`
but next code will not:
`int a = id<1>`

The specification of whole class is only way to specialize implicit conversion operator for one dimension.